flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGate.java

```diff
@@ -273,6 +273,8 @@ public int getNumberOfQueuedBuffers() {
 				for (InputChannel channel : inputChannels.values()) {
 					if (channel instanceof RemoteInputChannel) {
 						totalBuffers += ((RemoteInputChannel) channel).getNumberOfQueuedBuffers();
+					} else if (channel instanceof LocalInputChannel) {
```

Comment:
Maybe we could implement `getNumberOfQueuedBuffers` in parent `InputChannel` and returns default 0. For `RemoteInputChannel` and `LocalInputChannel` could override this method for returning real values. `UnknownInputChannel` do nothing. Then we could avoid `instanceof` here.

Reply:
A question here: At the aspect of metric method, should we use synchronized method to get the accurate number?  Because I see some method use the unsynchronized but still some directly work with synchronize like this `((RemoteInputChannel) channel).getNumberOfQueuedBuffers();` . So is there some principles on this @zhijiangW @pnowojski .

I'm in favour of working with synchronized, because the accurate metric will be more convictive . But I'm not sure about the performance impact. Waiting for you guys opinions. Thanks.

Reply:
I wouldn't change the current behaviour and contract, as that would require more extensive testing/benchmarking. Maybe in the future we might want to re think this, but I don't think this has a high priority right now.

Reply:
Agree! So i keep use the `unsynchronized` method style.